### PR TITLE
Server mode

### LINKED ISSUE #47: Can't connect to Client - is server mode possible?
Congratulations for the project and thank you for making it open source. 

Can ble-serial send a serial stream over BLE to an Android phone? 

I got your stack working as I am able to scan and ble-serial connect to an HM19 module. But instead to connect to an HM10/HM19, I would like to establish a serial link over BLE via a USB BLE dongle (without using the HM19) to my android phone.  

FYI without ble-serial and a USB-BLE dongle, I am able to connect my phone to the HM19: I first connected the HM19 to a USB-to-serial module CP210x and data sent to the COM port of the CP210x module would show on “Serial Bluetooth Terminal” on Android. 

How come my phone doesn’t show with a ble-scan?

Is It possible to ble-serial to the phone?

### PR DESCRIPTION
**Pull request by Jakeler**
closes #47

TODO:
- [x]  fix bugs with notify reordering
- [x] session handling
- [x] cherry pick commit from https://github.com/zalexua/ble-serial/commits/master/ regarding custom name
- [x] basic test on Windows
- [x] check async exceptions, especially under load
- [x] readme, mention multi device
- [x] check/update examples

### CODE REVIEW

**Comment by Jakeler** [comment]
Strange packet reordering issue, when receiving writes:
grafik
grafik
Needs more investigation before release.

**Comment by zalexua** [comment]
Hi !

I've read #47 , I saw there that it dies not work on Linux, where you referenced (kevincar/bless#60) which has been fixed now.

I've tried to test the "origin/server-mode" branch on Linux and it fails:
```
# ble-serial -d server -s 6e400001-b5a3-f393-e0a9-e50e24dcca9e -w 6e400002-b5a3-f393-e0a9-e50e24dcca9e -r 6e400003-b5a3-f393-e0a9-e50e24dcca9e -vv
01:35:33.490 | DEBUG | main.py: Running: Namespace(verbose=2, timeout=5.0, adapter='hci0', mtu=20, device='server', addr_type='public', service_uuid='6e400001-b5a3-f393-e0a9-e50e24dcca9e', write_uuid='6e400002-b5a3-f393-e0a9-e50e24dcca9e', read_uuid='6e400003-b5a3-f393-e0a9-e50e24dcca9e', mode='rw', filename=None, binlog=False, port='/tmp/ttyBLE', tcp_host='127.0.0.1', tcp_port=None)
01:35:33.490 | DEBUG | selector_events.py: Using selector: EpollSelector
01:35:33.491 | INFO | linux_pty.py: Port endpoint created on /tmp/ttyBLE -> /dev/pts/6
01:35:33.491 | INFO | ble_interface.py: Receiver set up
01:35:33.522 | DEBUG | manager.py: initial properties: {'/org/bluez': {..... I cut long output here manually .....}}
01:35:33.645 | DEBUG | manager.py: received D-Bus signal: org.freedesktop.DBus.Properties.PropertiesChanged (/org/bluez/hci0): ['org.bluez.Adapter1', {'Discovering': <dbus_fast.signature.Variant ('b', True)>}, []]
.... 5 seconds delay is here ....
01:35:38.767 | DEBUG | manager.py: received D-Bus signal: org.freedesktop.DBus.Properties.PropertiesChanged (/org/bluez/hci0): ['org.bluez.Adapter1', {'Discovering': <dbus_fast.signature.Variant ('b', False)>}, []]
01:35:38.767 | ERROR | main.py: Unexpected Error: AssertionError('No matching device found!')
01:35:38.767 | WARNING | main.py: Shutdown initiated
01:35:38.768 | INFO | linux_pty.py: Serial reader and symlink removed
01:35:38.768 | INFO | main.py: Shutdown complete.
```

Is this still supposed to be not working on Linux or there is some issue?

It looks like not trying to work as server, but actually waiting for device with name "server" or so.

I'd like to get server-mode working on Linux, so ready to work on that, but I need a guide to move on.

**Comment by zalexua** [comment]
Just FYI. I leave some comments in #47 about tests of this pool request.
I continue to provide updates there.

**Comment by Jakeler** [comment]
Retested with bless 0.2.6 and blueZ 5.79 and Py 3.12: packet reordering is not an issue anymore!
Ran >1MB through it at full speed client -> server with 255 mtu, received data is perfect.
Data is ok server -> client too, but it crashes silently after a while and then nothing is forwarded in either direction. On shutdown the async exception handler shows resource temporarily unavailable...

File: ble_serial/main.py [hunk 1]
```diff
@@ -42,25 +49,32 @@ async def _run(self):
...
             self.uart.start()
-            await self.bt.connect(args.device, args.addr_type, args.timeout)
-            await self.bt.setup_chars(args.write_uuid, args.read_uuid, args.mode, args.write_with_response)
+            
+            if args.gap_role == 'client':
+                await self.bt.connect(args.device, args.addr_type, args.timeout)
+                await self.bt.setup_chars(args.mode, args.write_with_response)
+            elif args.gap_role == 'server':
+                await self.bt.setup_chars(args.mode)
+                await self.bt.start(args.timeout)
```
Comment (Jakeler) [inline ble_serial/main.py:58]: Change with interface

**Comment by Jakeler** [comment]
Works fine on Windows too (without load testing):
grafik
grafik

Important for readme, needs manual install of this module which is not on pypi
```
pip install https://github.com/gwangyi/pysetupdi/archive/refs/heads/master.zip
```

**Comment by Jakeler** [comment]
Server is stable under normal conditions, also with repeated connect/disconnects now. No issues with 36k baud sustained. Sending from server works even with 80-100k baud, but not forever - here the mentioned silent crashes (11: resource temporarily unavailible) happen. Should be fine though for most apps and the release.

grafik
Also pytest is awesome, very happy with the new testing structure!

**Review by Jakeler — COMMENTED** [review]
LGTM

**Pull request merged**